controller/src/main/java/io/pravega/controller/store/stream/ZKScope.java

```diff
@@ -37,16 +72,207 @@ public String getName() {
 
     @Override
     public CompletableFuture<Void> deleteScope() {
-        return store.deleteNode(scopePath);
+        return store.deleteNode(scopePath)
+                    .thenCompose(v -> Futures.exceptionallyExpecting(store.deleteTree(counterPath), DATA_NOT_FOUND_PREDICATE, null))
+                    .thenCompose(v -> Futures.exceptionallyExpecting(store.deleteTree(streamsInScopePath), DATA_NOT_FOUND_PREDICATE, null));
+    }
+
+    CompletableFuture<Void> addStreamToScope(String name, int streamPosition) {
+        // break the path from stream position into three parts --> 2, 4, 4
+        String path = getPathForStreamPosition(name, streamPosition);
+        return Futures.toVoid(store.createZNodeIfNotExist(path));
+    }
+
+    CompletableFuture<Void> removeStreamFromScope(String name, int streamPosition) {
+        String path = getPathForStreamPosition(name, streamPosition);
+
+        return Futures.toVoid(store.deletePath(path, true));
+    }
+
+    private String getPathForStreamPosition(String name, int streamPosition) {
+        Preconditions.checkArgument(streamPosition >= 0);
+        Token token = new Token(streamPosition);
+
+        // root path represents most significant 2 digits of the position.
+        String root = ZKPaths.makePath(streamsInScopePath, token.getMsb().toString());
+        // container path represents most significant 3rd to 6th digits of the position.
+        String container = ZKPaths.makePath(root, token.getMiddle().toString());
+        // the path is composed of stream name followed by last 4 chars that represent least significant 4 digits of the position.
```

Comment:
Do we need this unique id for anything? Could we, for example, hash the stream name and use a prefix of the digits? It is fine if there is a collision on the hash because for pagination purposes, we are grouping streams anyway. It is a different story in the case you need the id for something else.

Reply:
we store this unique id for when we attempt to delete the stream, we should know where we will find it.